### PR TITLE
refactor: replace archiver with modern-tar

### PR DESCRIPTION
**Pull request by ayuhito**
## Overview

Related #1161 #1410

This replaces `archiver` with `modern-tar`. Nets in a reduced dependency surface and some performance improvements for related workloads (although most likely not noticeable).

There are likely some very low level differences between these libraries when it comes to error or symlink handling, but I would like to believe `modern-tar` follows more correct behaviour. `archiver` does a lot of non-standard things based on years of changes.

### Notes

Additionally, `modern-tar` is an ESM-only package, however, with dynamic imports we should be covered against cases like #1365.

I'll also make follow-up PRs to replace some instances of `tar-fs` in the codebase. Unfortunately, it's not possible to create stacked PRs from forks so that will happen after this PR is merged 📦

### CODE REVIEW

File: packages/testcontainers/src/utils/tar.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    ),
+    Promise.all(directories.map(createDirectorySources)),
+  ]);
+  const sources: TarSource[] = [...fileSources, ...directorySources.flat()];
+
+  // TAR headers need the content size up front, so buffer unknown-length streams before packing.
+  for (const { content, target, mode } of contents) {
+    sources.push({
+      type: "content",
+      content: content instanceof Readable ? await buffer(content) : content,
```
[the remark this answers is not in the document]
Comment (ayuhito) [reply packages/testcontainers/src/utils/tar.ts:36]: I was about to pre-comment this 🙈 `archiver` implicitly buffers `Readable`, where this is explicit. We do stream where we can elsewhere though!

https://github.com/testcontainers/testcontainers-node/pull/1161#discussion_r2429837291

**Comment by cristianrgreco** [comment]
> modern-tar is an ESM-only package

Ah, I didn't know modern-tar was ESM only 🫣 In that case this would require a major release, but I'd need to see if there are workarounds for Jest users. Although with dynamic imports we won't encounter issues at load time like in #1365, any _usage_ of modern-tar would still fail

**Comment by cristianrgreco** [comment]
I notice modern-tar has no deps, and uses tsdown for its build. Would it be possible / is it something you would want to do, to enable dual-publishing? I think it would be a config addition via tsdown, and shouldn't require any changes for existing users, but would add CJS compatibility. It would however increase your bundle size

**Comment by ayuhito** [comment]
Oops, yeah I didn't think that through. This PR can't be merged as is.

> enable dual-publishing?

In principle, I don't _want_ to support CJS as a way to push the ecosystem forward and also the added package bloat. Sorry for being a bit uncooperative on this matter. Jest is one of the unfortunate incumbents 😔 

> uses tsdown for its build

But, this actually is a really interesting idea if we were to apply it to `testcontainers-node`? If this package is open to using a bundler like `tsdown`, then it would be possible to inline bundle any future ESM dependencies as a workaround without needing a major release?

**Comment by cristianrgreco** [comment]
> In principle, I don't want to support CJS as a way to push the ecosystem forward and also the added package bloat. Sorry for being a bit uncooperative on this matter. Jest is one of the unfortunate incumbents 😔

No problem at all, I agree it would take away a bit the "modern" part of modern-tar 🙂 

> But, this actually is a really interesting idea if we were to apply it to testcontainers-node?

It is interesting. It'd solve such future problems as well. For example we're currently stuck on an old version of get-port for the same reason. 

There are downsides, e.g. our bundle size will increase, clients won't dedupe dependencies, etc. But for small libs, especially with no deps like modern-tar and get-port, it may be a good compromise. 

I will need time to look into it. A first approach may be to bundle the latest get-port and get that released. If it's all good then we could add modern-tar to the list in this PR and that'd be it

**Comment by ayuhito** [comment]
Feel free to close this PR if needed and handle things which you think is best! Happy to help in the future whenever 👍

**Comment by cristianrgreco** [comment]
Thanks @ayuhito. 

Btw what do you think if I were to create a fork of modern-tar which repackages modern-tar as CJS, and published it to a lib called `modern-tar-cjs`? I could setup Dependabot so that we automatically rebuild and republish any changes made to modern-tar. Could try my best to follow your versions, etc.

I am slightly hesitant to replace Testcontainer's build system, so wondering if managing forks is a better alternative

**Comment by ayuhito** [comment]
> create a fork

It would work but I fear it isn't an ideal precedent to set, nor would be a good burden to maintain.

> I am slightly hesitant to replace Testcontainer's build system, so wondering if managing forks is a better alternative

But I can understand this and the motivation for a fork.

I was curious what the `tsdown` migration could look like (relied on AI to churn on the edge cases) for a "best case" scenario. That includes enabling [package validation](https://tsdown.dev/options/lint) integrations like `publint` and `attw` as well as inlining `modern-tar`. The Jest smoke test now also adds file copying as part of the coverage. You can check it out here: https://github.com/ayuhito/testcontainers-node/commit/2678ee6dfbbe618a8edddca53e59a2280c02643c.

The diff is mostly `package-lock.json` (excluding it is `+284 / -695`). Builds are still outputted to `build/` but the shape is definitely not the same as `tsc`. Some postbuild scripts were moved to `tsdown.copy` too.

Not sure if this is the direction you want to go, but it does seem quite feasible?
